### PR TITLE
feat: Mongo Instrumentation (ported from Datadog)

### PR DESCRIPTION
**Pull request by johnnyshields**
Fixes https://github.com/open-telemetry/opentelemetry-ruby/issues/257

Ready for review

### CODE REVIEW

File: instrumentation/mongo/README.md
Comment (fbogsany) [inline instrumentation/mongo/README.md:3]:
```suggestion
The OpenTelemetry Mongo Ruby gem is a community maintained instrumentation for [Mongo][mongo-home].
```

File: instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb
Comment (fbogsany) [inline instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb:40]:
```suggestion
            span.set_attribute('net.peer.name', event.address.host)
            span.set_attribute('net.peer.port', event.address.port)
```

File: instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb
Comment (fbogsany) [inline instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb:35]:
```suggestion
            span.set_attribute('db.mongo.collection', collection) if collection
```

File: instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb
Comment (fbogsany) [inline instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb:31]:
```suggestion
```

File: instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb
Comment (fbogsany) [inline instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb:34]:
```suggestion
            span.set_attribute('db.operation', event.command_name)
```

File: instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb
Comment (fbogsany) [inline instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb:32]:
```suggestion
            span.set_attribute('db.system', 'mongodb')
```
This used to be `db.type` but appears to have changed to `db.system` at some point: https://github.com/open-telemetry/opentelemetry-specification/blob/323cd1d8811d3e3df08bec689f9dacd6cb782888/semantic_conventions/trace/database.yaml#L8

We will need to fix the Redis and MySQL instrumentation too.

File: instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb
Comment (fbogsany) [inline instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb:33]:
```suggestion
            span.set_attribute('db.name', event.database_name)
```

File: instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb
Comment (fbogsany) [inline instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb:38]:
```suggestion
```
These are not part of the OTel semantic conventions for Mongo: https://github.com/open-telemetry/opentelemetry-specification/blob/323cd1d8811d3e3df08bec689f9dacd6cb782888/semantic_conventions/trace/database.yaml#L283-L295

File: instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb
Comment (fbogsany) [inline instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb:49]: The error should be reported in a similar fashion to https://github.com/open-telemetry/opentelemetry-ruby/blob/576bc41f45dc0286a5c4a3c01bdac5efd6a27028/sdk/lib/opentelemetry/sdk/trace/span.rb#L125-L132

In this case, I think it should be something like:
```suggestion
              if event.is_a?(::Mongo::Monitoring::Event::CommandFailed)
                add_event('exception',
                          attributes: {
                            'exception.type' => 'CommandFailed',
                            'exception.message' => event.message
                          })
              end
```

File: instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb
Comment (fbogsany) [inline instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb:58]:
```suggestion
          def succeeded(event)
          end
```
This isn't part of the OpenTelemetry semantic conventions for Mongo.

File: instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb
Comment (fbogsany) [inline instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb:92]:
```suggestion
            Thread.current[THREAD_KEY]&.delete(event.request_id)
```

File: instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb
Comment (fbogsany) [inline instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb:26]:
```suggestion
            collection = get_collection(event.command)
            span = tracer.start_span("#{collection || ''}.#{event.command_name}", kind: :client)
```
Based on example at https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/database.md#mongodb

File: instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb
Comment (fbogsany) [inline instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb:30]:
```suggestion
```
Moved this up to set the span name.

File: instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb
Comment (johnnyshields) [reply instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb:38]: I think they should be... these will make it much easier to trace with respect to changes in the database, especially op_id which would allow the trace to be linked to the opLog

File: instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb
Comment (johnnyshields) [reply instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb:58]: Could we add this? This is the number of modified records. Again very useful to see in the trace and quickly explain latency.

**Comment by johnnyshields** [comment]
@fbogsany thank you so much for the code review. Would it be possible to consider extending the OpenTelemetry schema to incorporate:

```
db.mongo.request_id
db.mongo.op_id
db.mongo.n  (or count)
```

File: instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb
Comment (fbogsany) [reply instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb:35]: Could you please raise this as an issue (or open a PR) in the spec repo? I suspect they should be `db.mongodb.request_id` and `db.mongodb.op_id`, based on existing semantic conventions.

File: instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb
Comment (fbogsany) [reply instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb:57]: As above, could you please open an issue (or create a PR) for this in the spec repo? I think it'd require a prefix like `db.mongodb.` and should probably have a more meaningful name than `n`.

**Comment by fbogsany** [comment]
> @fbogsany thank you so much for the code review. Would it be possible to consider extending the OpenTelemetry schema to incorporate:
> 
> ```
> db.mongo.request_id
> db.mongo.op_id
> db.mongo.n  (or count)
> ```

It is possible. Please open a spec issue (or PR) making the case for these.

File: instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb
Comment (ericmustin) [reply instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb:57]: Ditto to @fbogsany comments here. I think it's a reasonable thing to add but i think it would gain broader support using a more descripitve naming convention

File: instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb
Comment (ericmustin) [inline instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb:25]: Generally I think this fits the specification per https://github.com/open-telemetry/opentelemetry-specification/blob/acd6f771d6064b92738a96afd729c59ba8ab9c4d/specification/trace/semantic_conventions/database.md#mongodb

That being said it looks like: 

- python has a slightly different approach here, with `mongodb.<collection>.<operation>` https://github.com/open-telemetry/opentelemetry-python/blob/f3cdfa2cdb9cb1c442189e2ead3788f45d92352d/instrumentation/opentelemetry-instrumentation-pymongo/src/opentelemetry/instrumentation/pymongo/__init__.py#L66

- and js is even less descriptive but also prepends the name with `mongodb` https://github.com/open-telemetry/opentelemetry-js-contrib/blob/088a4d57afa7a36c8f86c9afaa82d18477c82e5c/plugins/node/opentelemetry-plugin-mongodb/src/mongodb.ts#L232 

whereas ruby is just `<collection>.<operation>`. It seems like it's not part of the spec but it couldn't hurt  to prepend with `mongodb`

File: instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb
Comment (ericmustin) [inline instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb:15]: Broadly speaking I know that while the Datadog Implementation relies on thread local storage for it's mongo instrumentation, I'm not sure what Otel wants to support or if we have more 1st class helpers to do this. I'd like to get some other opinions here.

File: instrumentation/mongo/test/test_helper.rb
Comment (ericmustin) [inline instrumentation/mongo/test/test_helper.rb:48]: I believe the datadog test approach relies on a separate mongodb container in the docker network. I think we'd need to add this env var pointing to the appropriate container in our docker-compose.yml https://github.com/open-telemetry/opentelemetry-ruby/blob/master/docker-compose.yml and add both the mongodb container and the relevant env vars to the ci.yml https://github.com/open-telemetry/opentelemetry-ruby/blob/master/.github/workflows/ci.yml

**Review by ericmustin — CHANGES_REQUESTED** [review]
I think this broadly lgtm, left a few comments.

- there's a few outstanding items to do to make sure the test suite can run locally and in CI
- I'd like some more experienced eyes on the use of thread local storage here. it's possible there's a more canonical way to handle this within the otel api or sdk 
- a few minor nits on the attribute metadata naming and details

File: instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb
Comment (johnnyshields) [reply instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb:25]: I agree to prepend with mongodb to be consistent with python.

**Comment by johnnyshields** [comment]
@ericmustin re: "use of thread local storage", I don't see any other way clean around this besides to use thread storage. The mongo-ruby-driver's instrumentation API gives us separate "start" and "end" callbacks, so we need to store the ID somewhere during the time between. They don't give us the equivalent of an "around (with yield)" callback. The only real alternative would be to hack-in our own monkey patch instrumentation which is ill-advised...

**Comment by ericmustin** [comment]
@johnnyshields yup I think thread local is the way to go, I just mean that if we have some existing helpers within the API for managing this we should use those. I don't believe we do though, I think your implementation is good on this front.

File: instrumentation/mongo/test/test_helper.rb
Comment (johnnyshields) [reply instrumentation/mongo/test/test_helper.rb:48]: The CI passes fine here as-is. If you want to run this on a different CI, you will need to set `TEST_MONGODB_HOST` and `TEST_MONGODB_PORT` in that CI's runner script. I'm marking this as resolved as I don't see any code change is needed.

**Comment by johnnyshields** [comment]
@ericmustin @fbogsany I believe I've addressed all comments. Can you please review again?

File: instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb
Comment (johnnyshields) [reply instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb:15]: Following discussion below, this should be acceptable for now, and if in the future OTEL ruby provides an API for thread storage we can refactor in a separate PR. Resolving discussion.

File: instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb
Comment (fbogsany) [inline instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb:66]:
```suggestion
          def span_name(collection, command_name)
            return command_name unless collection
            "#{collection}.#{command_name}"
          end
```
Despite what other languages have done, we should strive to follow the semantic conventions where possible. Deciding **not** to do that leads to further divergence rather than convergence.

File: instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb
Comment (fbogsany) [reply instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb:15]: Ideally, we'd use `Context` for this, but that tends to assume consistent nesting of operations. I this case, it looks like we can't really avoid the thread-local.

File: instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb
Comment (ericmustin) [reply instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb]: my bad @johnnyshields . Agreed, we should follow [The Spec](https://youtu.be/o31UcPeiUCc)

**Review by ericmustin — APPROVED** [review]
nice work here, i think everything has been addressed afaik

File: instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb
Comment (johnnyshields) [reply instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/middlewares/subscriber.rb:15]: @fbogsany yep originally I actually tried using Context but then abandoned it.

**Comment by johnnyshields** [comment]
Let's merge?

**Comment by johnnyshields** [comment]
@mwear @dazuma please approve

File: instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/instrumentation.rb
Comment (mwear) [inline instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/instrumentation.rb:33]: It might be more accurate to call this `register_subscriber` or something along those lines.

**Review by mwear — APPROVED** [review]
LGTM. One minor comment.

File: instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/instrumentation.rb
Comment (fbogsany) [reply instrumentation/mongo/lib/opentelemetry/instrumentation/mongo/instrumentation.rb:33]: It's also more accurate to call the `Middlewares` module something like `Subscribers`, or perhaps remove that module altogether and move `Subscriber` and friend to the `Mongo` module.

**Review by fbogsany — APPROVED** [review]

**Comment by fbogsany** [comment]
LGTM after the CI failure is fixed.

**Comment by johnnyshields** [comment]
@fbogsany specs fixed lets merge

**Comment by fbogsany** [comment]
Thanks @johnnyshields. This needs a rebase before I can merge. I don't have permissions to rebase - can you please do that?

**Review by robertlaurin — APPROVED** [review]

**Comment by johnnyshields** [comment]
Merged in master

**Comment by johnnyshields** [comment]
Ready to merge

**Comment by fbogsany** [comment]
Thanks @johnnyshields

**Pull request merged**
